src/main/scala/lms/core/stub.scala

```diff
@@ -80,6 +80,38 @@ object Adapter extends FrontEnd {
 
   class MyGraphBuilder extends GraphBuilder {
 
+    override def gatherEffectDeps(reads: Set[Exp], writes: Set[Exp], s: String, as: Def*): (Set[Sym], Set[Sym]) = {
+      val (prevHard, prevSoft) = (new mutable.ListBuffer[Sym], new mutable.ListBuffer[Sym])
+      // gather effect dependencies 1): handle the write keys
+      writes.foreach(key => curEffects.get(key) match {
+        case Some((lw, lr)) =>
+          // Fine Grained HardDeps:
+          val lastWrite = latest(lw)
+          if (s == "array_set") {
+            findDefinition(lastWrite) match {
+              // If latest(lw) is on the same array at the same index, we do not add hard dependence but soft dependence
+              case Some(Node(_, "array_set", List(as2, idx2, value2), deps)) if (as(0) == as2 && as(1) == idx2) =>
+                prevSoft += lastWrite
+                prevSoft ++= deps.sdeps
+                prevHard ++= deps.hdeps
+              case _ => prevHard += lastWrite
+            }
+          } else {
+            prevHard += lastWrite;
+          }
+          // write has soft deps to previous reads
+          if (!reflectHere) prevSoft ++= lr
+        case _ =>
+          // write has hard dependencies on declaration (if declared locally) or block (if declared globally, i.e., out of current block)
+          prevHard += latest(key);
+      })
+      // gather effect dependencies 2): if we enforce reifyHere (reify in current block), add curBlock as hard dependencies
+      if (reifyHere) prevHard += curBlock
+      // gather effect dependencies 3): handle read keys (i.e., reads have hard dependencies on previous write)
+      reads.foreach(key => prevHard += getLastWrite(key))
+      (prevHard.toSet, prevSoft.toSet)
+    }
+
```

Comment:
this function overwrites the `gatherEffectDeps` case and check for finer-grained write keys and deps

Reply:
I think it is good to have an override, but that would potentially break everything. The whole logic has to be rewritten.
I don't know if the backend should give that much control to the front end.


Can we make it composable somehow? I didn't think of a solution, but giving something like:

given a list of (cond, soft, hard)

cond: (s: Sym, n: Node) => Boolean, condition to apply the transformation:
```
s == "array" && n == Node(_, "array_set", List(as2, idx2, value2), deps)) && (as(0) == as2 && as(1) == idx2)
```
soft: Node => Set[Sym]
hard: Node => Set[Sym]

I don't know if it is generic enough. But in the case the node doesn't match any condition, use the generic one.

Reply:
yeah. I think doing `case` with `falling back to the generic case` is much better :) Will push a fix soon

Reply:
fixed